Adding process_immutable_with_scratch method

michaelciraci opened the pull request:
Follow on for #150 

I found it easier to follow up https://github.com/ejmahler/RustFFT/pull/151#issuecomment-2709159596 with a fresh PR, so this is instead of https://github.com/ejmahler/RustFFT/pull/151.

I implemented the immutable method for scalar, AVX, SSE, Neon, and wasm, so no default method was actually needed.

ejmahler commented on src/algorithm/bluesteins_algorithm.rs:
```diff
@@ -137,6 +137,49 @@ impl<T: FftNum> BluesteinsAlgorithm<T> {
+        }
+    }
+
     fn perform_fft_out_of_place(
```
One thing I'd like to see here is that, since the out of place FFT is identical to the immutable FFT in every way, this could just forward to the immutable FFT to eliminate some code duplication.

michaelciraci replied:
Yes, there's several places this happens. To remove the extra function call I'm in favor of marking the out of place as inline if that works with you.

ejmahler commented on src/algorithm/mixed_radix.rs:
```diff
@@ -151,6 +151,47 @@ impl<T: FftNum> MixedRadix<T> {
+        // process_with_scratch can transpose the output into the input variable, saving scratch for just fft scratch
+        // Since we can't use the input variable, allocate twice as much scratch as process_with_scratch,
+        // and split them into two scratches we can use
+        let (scratch, scratch2) = scratch_raw.split_at_mut(scratch_raw.len() / 2);
```
Unfortunately, this logic isn't sufficient to cover all of our use cases. Check out this test to see why:

Essentially, this code assumes that the scratch the implace transform requested is exactly equal to the FFT length, but you can see [here](https://github.com/ejmahler/RustFFT/blob/c7a81dad8223dc9f48fb8b7c2db458a986d06bec/src/algorithm/mixed_radix.rs#L96) that that isn't true.

I recommend adding this function to [this](https://github.com/ejmahler/RustFFT/blob/c7a81dad8223dc9f48fb8b7c2db458a986d06bec/src/algorithm/mixed_radix.rs#L419) test, which intentionally stress-tests the logic of how much scratch is requested, and how it gets distributed around. THere's also one for the good-thomas algorithm

michaelciraci replied:
I updated the immut scratch logic, and added the testing for both mixed radix and good-thomas. Let me know if that's closer to what you had in mind.

ejmahler commented:
It's great to see so much progress! I think this is close, but there are still a few things we need to work on. As I mentioned in my comment above, the biggest thing is to just make sure that the amount of scratch we're requesting is sound. I know from experience that it's difficult to wrap my head around and easy to get wrong, so it's worth paying extra attention to it.

If you want a specific example where this will go wrong with Mixed Radix and Good-Thomas in this PR, try constructing this in a local test: A MixedRadix instance (or good-thomas instance), with the children being a Butterfly7 on one side, and a BluesteinsAlgorithm with a length of 100 and inner FFT of 256 on the other side, and then compute an immutable-input FFT. The current code will not distribute scratch correctly and I bet it will panic.

This is also making me realize that Radix4 could use some more testing. Last year, I added a PR to allow Radix4 to take arbitrary FFTs as the base, which is a fantastic expansion of functionality, but opens the door for similar scratch bugs like this. I should have added the big scratch test I linked above, but it didn't occur to me.

Since scratch size is a concern again, especially with this new code, would you mind implementing big scratch tests for the various Radix4 implementations, similar to the one that exists for MixedRadix? That way we could get code coverage on this new pathway's scratch logic for Radix4.

ejmahler commented:
One more thing: The `[check_fft_algorithm](https://github.com/ejmahler/RustFFT/blob/c7a81dad8223dc9f48fb8b7c2db458a986d06bec/src/test_utils.rs#L70)` method should have a new section at the bottom that tests the new immutable FFT method. I imagine it can largely be copied from the out of place version, just with immutable input.

michaelciraci commented:
Sounds good. I was leaving it up to the planner to chose an algorithm for each length, but I'll use that method for coverage for all the different algorithms.

ejmahler commented on src/test_utils.rs:
```diff
@@ -173,6 +173,39 @@ pub fn check_fft_algorithm<T: FftNum + Float + SampleUniform>(
+
+            assert!(
+                compare_vectors(&expected_output, &output),
+                "process_outofplace_with_scratch() failed the 'dirty scratch' test, length = {}, direction = {}",
```
Minor: The assert failure message needs to be updated.

michaelciraci commented:
Corrected the scalar implementation, and tested the permutation of all features. Pipeline should pass now.

michaelciraci commented:
Also updated benches.

michaelciraci commented:
@ejmahler Just a friendly follow up. Should be good :)

ejmahler commented:
Thanks for waiting. Since it's so big, I was waiting for some free time to do a more thorough review, and I have some now.

ejmahler commented on src/algorithm/dft.rs:
```diff
@@ -45,6 +45,15 @@ impl<T: FftNum> Dft<T> {
+        spectrum: &mut [Complex<T>],
+        _scratch: &mut [Complex<T>],
+    ) {
+        self.perform_fft_out_of_place(signal, spectrum, _scratch);
```
Minor feedback: For consistency's sake, I'd argue that we should swap these, so that the out of place version calls the immutable version.

ejmahler commented on src/algorithm/good_thomas_algorithm.rs:
```diff
@@ -241,6 +250,30 @@ impl<T: FftNum> GoodThomasAlgorithm<T> {
         self.reindex_output(scratch, buffer);
     }
 
+    fn perform_fft_immut(
```
I think the scratch usage can be simplified here.

```rust
fn perform_fft_immut(
        &self,
        input: &[Complex<T>],
        output: &mut [Complex<T>],
        scratch: &mut [Complex<T>],
    ) {
        // Re-index the input, copying from the input to the output in the process
        self.reindex_input(input, output);

        // run FFTs of size `width`
        self.width_size_fft.process_with_scratch(output, scratch);

         let (scratch, inner_scratch) = scratch.split_at_mut(self.len());

        // transpose
        transpose::transpose(output, scratch, self.width, self.height);

        // run FFTs of size 'height'
        self.height_size_fft.process_with_scratch(scratch, inner_scratch);

        // Re-index the output, copying from the input to the output in the process
        self.reindex_output(scratch, output);
    }
```

From there, we can analyze the amount of scratch required: We pass the whole thing unmodified to the width-size FFT, so our immutable scratch must be at least as large as our width FFT in-place scratch. Then, we take `self.len()` scratch for ourselves and send the rest to the height-size FFT, so our immutable scratch must be at least as large as our height-size FFT in-place scratch plus `self.len(). Thus, the immutable scratch required on line 123 is `max(width_fft.get_inplace_scratch_len(), self.len() + height_fft.get_inplace_scratch_len()`.

ejmahler commented on src/algorithm/good_thomas_algorithm.rs:
```diff
@@ -532,12 +599,15 @@ mod unit_tests {
+                            len,
+                            inplace_scratch,
+                            outofplace_scratch,
+                            direction: FftDirection::Forward,
```
Formatting nitpick: `direction` should be on the bottom here so the in place scratch can be grouped with the others.

ejmahler commented on src/algorithm/mixed_radix.rs:
```diff
@@ -103,6 +104,14 @@ impl<T: FftNum> MixedRadix<T> {
+        let height_fft_immut = height_fft.get_immutable_scratch_len();
+        let width_fft_immut = width_fft.get_immutable_scratch_len();
+
+        let immut_scratch_len = 2 * max(
```
Same scratch size analysis here as with good-thomas algorithm

ejmahler replied:
Although now that I read this more closely, it looks like mixed radix swaps width and height compared to mixed radix, so those will have to be swapped here too.

ejmahler commented on src/algorithm/mixed_radix.rs:
```diff
@@ -151,6 +161,47 @@ impl<T: FftNum> MixedRadix<T> {
+        // process_with_scratch can transpose the output into the input variable, saving scratch for just fft scratch
+        // Since we can't use the input variable, allocate twice as much scratch as process_with_scratch,
+        // and split them into two scratches we can use
+        let (scratch, scratch2) = scratch_raw.split_at_mut(scratch_raw.len() / 2);
```
Same revised scratch-tracking algorithm as I wrote for good-thomas algorithm. Pass the whole thing to the height FFT, then split_at_mut(self.len()), transpose into the left scratch, and pass the right scratch to the width FFT.

ejmahler commented on src/algorithm/raders_algorithm.rs:
```diff
@@ -115,10 +119,63 @@ impl<T: FftNum> RadersAlgorithm<T> {
+        // The first output element is just the sum of all the input elements, and we need to store off the first input value
+        let (output_first, output) = output.split_first_mut().unwrap();
+        let (input_first, input) = input.split_first().unwrap();
+        let (scratch, scratch2) = scratch.split_at_mut(output.len());
```
I think the scratch handling can be simplified here. First, I'd do this:

`let (scratch, extra_scratch) = scratch.split_at_mut(self.len() - 1);`

Then pass extra_scratch to the inner FFT on line 147 like this: `self.inner_fft.process_with_scratch(scratch, extra_scratch );`.

On line 157, keep the inner FFT data inside the buffer like the in-place FFT does. Then on line 169, call the inner FFT exactly as before: `self.inner_fft.process_with_scratch(scratch, extra_scratch );`

Then on line 173, copy from scratch instead of scratch2.

All of this will simplify the scratch analysis on line 106 to simply be `inner_fft_len + extra_inner_scratch` IE exactly the same as the in-place FFT.

michaelciraci replied:
I agree with the simplification of scratch usage, but I believe the scratch amount should actually be `inner_fft_len + required_inner_scratch`.

Just as an example, for `len` = 3:
`inner_fft_len` = 2
`required_inner_scratch` = 2

In the Raders new method, `extra_inner_scratch` turns out to be 0 because `required_inner_scratch <= inner_fft_len`.

Then when we get down to the perform, we have this line: `self.inner_fft.process_with_scratch(scratch, extra_scratch)` which fails because we have no scratch.

Setting the immutable scratch length to be `inner_fft_len + required_inner_scratch` works great, but I'll wait for your response on this just to make sure you agree.

ejmahler replied:
You're right, I overlooked what `extra_inner_scratch` contained, `required_inner_scratch` is the right one to use.

ejmahler commented on src/algorithm/radix3.rs:
```diff
@@ -153,6 +153,15 @@ impl<T: FftNum> Radix3<T> {
         }
     }
+
+    fn perform_fft_out_of_place(
```
IMO radix3, radix4, and radixn should all be consistent here: Either all of their OOP should just forward to the immutable version, or all of their OOP should have a dedicated implementation. Since radix4 and radixn have dedicated implementations, I vote for giving this its dedicated implementation back.

ejmahler commented on src/avx/avx32_butterflies.rs:
```diff
@@ -5,8 +5,8 @@ use std::mem::MaybeUninit;
 use crate::array_utils;
-use crate::array_utils::workaround_transmute_mut;
 use crate::array_utils::DoubleBuf;
+use crate::array_utils::*;
```
I'm personally not a fan of * imports, because it obfuscates symbols' origins. Please manually list out the symbols being imported here.

ejmahler commented on src/avx/avx32_butterflies.rs:
```diff
@@ -37,9 +37,9 @@ macro_rules! boilerplate_fft_simd_butterfly {
+                input: &[Complex<T>],
                 output: &mut [Complex<T>],
                 _scratch: &mut [Complex<T>],
             ) {
```
One pitfall here is that `fft_error_outofplace` has a panic message that specifically mentions the method name `process_outofplace_with_scratch`, so using that from here will give a confusing error message.

Because of that, I sadly don't think the butterfly boilerplate macros can forward their implementations like this. I think we should make a new `fft_error_immut` function that mentions the correct function name, and call it from here, and provide a full out of place implementation that calls its own method. This goes for all of the butterfly implementations.

ejmahler commented on src/avx/avx64_butterflies.rs:
```diff
@@ -155,6 +167,22 @@ macro_rules! boilerplate_fft_simd_butterfly_with_scratch {
+
             #[inline]
             fn perform_fft_out_of_place(
                 &self,
```
This can should forward to the immutable version.

ejmahler commented on src/avx/avx_bluesteins.rs:
```diff
@@ -311,6 +312,61 @@ impl<A: AvxNum, T: FftNum> BluesteinsAvx<A, T> {
+
     fn perform_fft_out_of_place(
         &self,
         input: &[Complex<T>],
```
This should forward to the immutable version

ejmahler commented on src/avx/avx_mixed_radix.rs:
```diff
@@ -862,6 +1039,125 @@ impl<A: AvxNum, T: FftNum> MixedRadix16xnAvx<A, T> {
+            for c in 0..self.len() / len_per_row {
+                let cs = c * len_per_row + len_per_row - partial_remainder;
+                match partial_remainder {
+                    1 => buffer.store_partial1_complex(input.load_partial1_complex(cs), cs),
```
Is it feasible to skip this and load directly from `input` down below?

ejmahler commented on src/avx/avx_mixed_radix.rs:
```diff
@@ -152,6 +192,143 @@ macro_rules! mixedradix_gen_data {
+            // If we don't have any column FFTs, we need to move the
+            // input into the output buffer for the remainder processing
+            if chunk_count == 0 {
+                let simd_ops = self.len() / 4;
```
After stepping through some tests,, it appears to pass because the loop on line 252 does a full copy of the data. So at minimum this branch is superfluous and should be removed. But imo the loop on line 252 should also be removed and the remainder code should load directly from the input buffer.

michaelciraci replied:
Yes, this did end up being superfluous. I forget why I added this, but it works for all lengths when I remove this.

ejmahler commented on src/avx/avx_raders.rs:
```diff
@@ -278,8 +280,9 @@ impl<A: AvxNum, T: FftNum> RadersAvx2<A, T> {
-            inplace_scratch_len: len + extra_inner_scratch,
+            inplace_scratch_len,
             outofplace_scratch_len: extra_inner_scratch,
+            immut_scratch_len: inplace_scratch_len * 2,
```
Same immutable scratch length analysis as the scalar version

michaelciraci replied:
I am unsure this is the same as the scalar. I think we actually need the scalar version + 1:

Using `RadersAvx2` for `len` = 3, if we were to match the scalar version, this means `inner_fft_len` = 2, and `required_inner_scratch` = 2 (`immut_scratch_len` = 4).

This puts both `scratch` and `extra_scratch` at `len` = 2. In the `out_of_place` version, the transmuted slices are of length 3: https://github.com/michaelciraci/RustFFT/blob/make-input-immutable-4/src/avx/avx_raders.rs#L476-L482

michaelciraci replied:
(I committed the scalar + 1 version if you'd like to take a look at it)

ejmahler commented on src/avx/avx_raders.rs:
```diff
@@ -351,6 +354,66 @@ impl<A: AvxNum, T: FftNum> RadersAvx2<A, T> {
+        let (first_input, _) = input.split_first().unwrap();
+        let (first_output, inner_output) = output.split_first_mut().unwrap();
+
+        let (scratch, scratch2) = scratch.split_at_mut(scratch.len() / 2);
```
Same scratch handling analysis as the scalar version

ejmahler commented on src/avx/mod.rs:
```diff
@@ -23,13 +23,57 @@ struct CommonSimdData<T, V> {
+                    || output.len() != input.len()
+                {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(
```
Should be rerouted to our soon-to-be-created `fft_error_immut` method, as with the one on line 67

ejmahler commented on src/avx/mod.rs:
```diff
@@ -130,6 +178,52 @@ macro_rules! boilerplate_avx_fft {
+                    || output.len() != input.len()
+                {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(
```
`fft_error_immut`

ejmahler commented on src/common.rs:
```diff
@@ -73,6 +73,45 @@ pub fn fft_error_outofplace(
+                    || scratch.len() < required_scratch
+                {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(
```
`fft_error_immut`

ejmahler commented on src/common.rs:
```diff
@@ -156,6 +195,10 @@ macro_rules! boilerplate_fft_oop {
             }
+            #[inline(always)]
+            fn get_immutable_scratch_len(&self) -> usize {
+                self.inplace_scratch_len()
```
I don't think this is correct. There are four uses of this macro - Dft, Radix3, Radix4, RadixN

For Dft, it's just 0. For Radix#, it's `inner_fft.len()`. IMO this should call `self.immut_scratch_len()` so we can customize these.

ejmahler replied:
Yes, that's perfect

ejmahler commented on src/common.rs:
```diff
@@ -173,8 +216,49 @@ macro_rules! boilerplate_fft_oop {
+                    || input.len() < self.len()
+                    || output.len() != input.len()
+                {
+                    fft_error_outofplace(
```
`fft_error_immut`

ejmahler commented on src/neon/neon_butterflies.rs:
```diff
@@ -82,6 +82,36 @@ macro_rules! boilerplate_fft_neon_f32_butterfly {
                 }
```
Out of place version should forward to immutable version.

ejmahler replied:
Actually I have a different recommendation. There's an incisistency between this and the SSE implementation, and imo the SSE implementation does it a better way. In that one, in the f32 and f64 macros there is no distinction between immutable and out of place versions, it's just the same method call. And the common implementations on line 194 and 213 just both call the same `perform_oop_fft_butterfly_multi` method

ejmahler commented on src/neon/neon_butterflies.rs:
```diff
@@ -112,9 +142,29 @@ macro_rules! boilerplate_fft_neon_f64_butterfly {
                 input: &mut [Complex<T>],
                 output: &mut [Complex<T>],
+            ) -> Result<(), ()> {
+                array_utils::iter_chunks_zipped_mut(
```
Should forward to immutable version

ejmahler replied:
Updated recommendation is the same as above, `perform_oop_fft_butterfly_multi` should be the only method here

ejmahler commented on src/neon/neon_butterflies.rs:
```diff
@@ -130,6 +180,25 @@ macro_rules! boilerplate_fft_neon_f64_butterfly {
+            ) {
+                if input.len() < self.len() || output.len() != input.len() {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(self.len(), input.len(), output.len(), 0, 0);
```
`fft_error_immut`

ejmahler commented on src/neon/neon_common.rs:
```diff
@@ -57,6 +57,37 @@ macro_rules! separate_interleaved_complex_f32 {
+
+                if input.len() < self.len() || output.len() != input.len() {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(self.len(), input.len(), output.len(), 0, 0);
```
`fft_error_immut`

ejmahler commented on src/neon/neon_radix4.rs:
```diff
@@ -83,6 +83,49 @@ impl<N: NeonNum, T: FftNum> NeonRadix4<N, T> {
+
     //#[target_feature(enable = "neon")]
     unsafe fn perform_fft_out_of_place(
         &self,
```
Out of place version should forward to immutable version

ejmahler commented on src/sse/sse_butterflies.rs:
```diff
@@ -176,6 +176,25 @@ macro_rules! boilerplate_fft_sse_f64_butterfly {
+            ) {
+                if input.len() < self.len() || output.len() != input.len() {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(self.len(), input.len(), output.len(), 0, 0);
```
fft_error_immut

ejmahler commented on src/sse/sse_common.rs:
```diff
@@ -90,6 +90,14 @@ macro_rules! boilerplate_fft_sse_oop {
+                output: &mut [Complex<T>],
+                scratch: &mut [Complex<T>],
+            ) {
+                self.process_immutable_with_scratch(input, output, scratch);
```
Unfortunately can't forward inside the boilerplante macros because of the error message thing

ejmahler commented on src/wasm_simd/wasm_simd_butterflies.rs:
```diff
@@ -134,6 +134,25 @@ macro_rules! boilerplate_fft_wasm_simd_f64_butterfly {
+            ) {
+                if input.len() < self.len() || output.len() != input.len() {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(self.len(), input.len(), output.len(), 0, 0);
```
`fft_error_immut`

ejmahler commented on src/wasm_simd/wasm_simd_common.rs:
```diff
@@ -57,6 +57,37 @@ macro_rules! separate_interleaved_complex_f32 {
+
+                if input.len() < self.len() || output.len() != input.len() {
+                    // We want to trigger a panic, but we want to avoid doing it in this function to reduce code size, so call a function marked cold and inline(never) that will do it for us
+                    fft_error_outofplace(self.len(), input.len(), output.len(), 0, 0);
```
`fft_error_immut`

ejmahler commented on src/wasm_simd/wasm_simd_radix4.rs:
```diff
@@ -82,6 +82,50 @@ impl<S: WasmNum, T: FftNum> WasmSimdRadix4<S, T> {
+
     #[target_feature(enable = "simd128")]
     unsafe fn perform_fft_out_of_place(
         &self,
```
Out of place version should forward to immutable version

ejmahler commented:
I left a whole bunch of comments. This is looking great! Thanks for putting so much work into this. it's much closer than before.

There are a few broad categories under which most of my feedback fell:

- Optimization of scratch handling, and optimization of amount of scratch requested. I wrote detailed instructions for each case.
- Something I overlooked before is that `fft_error_outofplace` mentions the outofplace FFT method by name in its panic message, so we will need to make a new one for the immutable case. This means we can't forward the `Fft` trait impls of out of place directly to the immutable impls, sadly.
- Many fft bodies can be forwarded from the out of place version to the immutable version

One thing I noticed is that there is no radix4 implementation for SSE. Is that intentional, or was it accidentally omitted?

I'm pretty sure I'll be ready to merge once all this is addressed. Thanks again for working on this.

michaelciraci commented on src/avx/avx_mixed_radix.rs:
```diff
@@ -69,6 +69,44 @@ macro_rules! boilerplate_mixedradix {
+        ) {
+            // Perform the column FFTs
+            // Safety: self.perform_column_butterflies() requires the "avx" and "fma" instruction sets, and we return Err() in our constructor if the instructions aren't available
+            let (scratch, scratch2) = scratch.split_at_mut(scratch.len() / 2);
```
I believe the scratch should be `let (scratch, inner_scratch) = scratch.split_at_mut(input.len());`, but otherwise this is addressed

ejmahler commented:
Regarding the autogenerated code match failure, it looks like rustfmt ran on a bunch of ignored files, including our benchmarks and autogenerated prime butterfly code. I recommend reverting those.

michaelciraci commented:
Also, I know I'm adding multiple commits. I'll squash commits before you merge it in.

michaelciraci commented:
> One thing I noticed is that there is no radix4 implementation for SSE. Is that intentional, or was it accidentally omitted?

Right now, the `immut` actually matches the `out_of_place` implementation, so there's no separate function: https://github.com/ejmahler/RustFFT/blob/0620130ab931d724dc1c1f5cacc7f99f984cfb1d/src/sse/sse_common.rs#L60-L100

I can reverse the call order on this one as well if you'd like the `out_of_place` to call `immut`

michaelciraci commented:
Neon and wasm are the same way where they don't need a specific immutable implementation. AVX requires a specific immutable implementation.

So if you want AVX, SSE, Neon, and wasm all consistent I could add an SSE immutable method, but that would simply just forward to out of place. If you don't care if all SIMD is consistent, I can remove the Neon and wasm forwarding.

HEnquist commented:
> if the SSE one doesn't need a dedicated immutable implementation, do neon and wasm simd? or can they be kept consistent with sse?

The neon, wasm and SSE implementations are closely related, IMO they should get the same treatment here as far as possible.

ejmahler commented:
Yeah. The AVX implementation was written earlier with a different, more experimental style, so I'm fine with it being different. The neon, wasm, and sse implementations are almost identical, and it'll keep them more maintainable if we keep them that way.

michaelciraci commented:
Corrected the wasm imports and bench `rustfmt`

michaelciraci commented:
Agh, accidentally had the wrong default when I added that macro.

I also updated the autogen test if you can run the pipeline again. Sorry about that...

michaelciraci commented:
Actually, I think we're going to have an issue with the autogen check because SSE and Neon have different imports

ejmahler commented:
If you're having trouble with the workflow of the autogenerated files, the way is works is that you edit the template, and then you run the autogen for sse, neon, and wasm simd.

I recommend adding fft_error_immut back to the list of imports in the template, then run the script for all 3 files. Then you can use it in the macro the same way everything else does.

ejmahler commented:
I'm not sure what's going on with the wasm simd error. IF you need help with that, I'll investigate that myself later tonight and push a fix myself once I figure out what's going on.

michaelciraci commented:
If you don't mind taking a look, I'd definitely accept the offer.

michaelciraci commented:
All the changes look good and make sense to me.

ejmahler commented:
Excellent! You put a ton of work into this, and I really appreciate it.

I have some small refactoring I want to do before making a release, which I will be able to do this weekend, so the release will probably be at the end of the weekend or next week.

michaelciraci commented:
Sounds good.

And likewise, thanks for your assistance with during this merge request.
